htlcinterceptor/middleware.go

```diff
@@ -0,0 +1,146 @@
+package htlcinterceptor
+
+import (
+	"errors"
+	"sync"
+	"sync/atomic"
+
+	"github.com/lightningnetwork/lnd/channeldb"
+	"github.com/lightningnetwork/lnd/lntypes"
+	"github.com/lightningnetwork/lnd/lnwire"
+)
+
+var (
+	//ErrFwdNotExists is an error returned when the caller tries to resolve
+	// a forward that doesn't exist anymore.
+	ErrFwdNotExists = errors.New("forward does not exist")
+)
+
+// Middleware is a middleware for other sub systems, including the RPC layer,
+// to be able to intercept htlc forwards. It holds a list of handlers that are
+// invoked sequentially per forward to figure out if one of them is interested
+// to handle the forward outside the switch. In addition it exposed API for
+// resolving existing hold forward.
+type Middleware struct {
+	sync.RWMutex
+	// ForwardHtlcHandler is the callback that is called for each forward of
+	// incoming htlc. It should return true if it is interested in handling
+	// the forwardHtlc.
+	fwdHandlers  map[uint64]ForwardHtlcHandler
+	fwdHandlerID uint64
+
+	// holdForwards is a map of current hold forwards and their corresponding
+	// ForwardResolver.
+	holdForwards map[channeldb.CircuitKey]ForwardResolver
+}
+
+// NewMiddleware initializes a Middleware.
+func NewMiddleware() *Middleware {
+	return &Middleware{
+		fwdHandlers:  make(map[uint64]ForwardHtlcHandler),
+		holdForwards: make(map[channeldb.CircuitKey]ForwardResolver),
+	}
+}
+
+// AddForwardHtlcHandler adds a ForwardHtlcHandler.
+func (c *Middleware) AddForwardHtlcHandler(
+	handler ForwardHtlcHandler) uint64 {
+
+	id := atomic.AddUint64(&c.fwdHandlerID, 1)
```

Comment:
do you need an atomic here when `fwdHandlers` is always accessed with a lock?

Reply:
No, moved to within the lock scope.